### PR TITLE
Graphing Device Dependency

### PR DESCRIPTION
**Pull request by SourceDoctor**
Graphing Device Dependency

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by kkrumm1** [comment]
Can you post screenshot of this?

**Comment by SourceDoctor** [comment]
@kkrumm1 @PipoCanaja 
just a reuse of "Communication Map" Graphiz Graph ;-)

Lonesome Nodes have no dependencies. They are graphed also so it can be seen they are "alone".

image

File: includes/html/print-dependency-map.inc.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
+<?php
+
+use LibreNMS\Config;
+
+if (!Auth::user()->hasGlobalRead()) {
+    $sql = 'WHERE device_id IN (SELECT device_id FROM devices_perms WHERE user_id='. Auth::id() . ')';
```
Comment (Jellyfrog) [inline includes/html/print-dependency-map.inc.php]: Please use Permission class to get user devices

**Comment by Jellyfrog** [comment]
This page could be written with blade/laravel, see #10559 for an example

File: includes/html/print-dependency-map.inc.php (hunk 1 above)
Comment (SourceDoctor) [reply includes/html/print-dependency-map.inc.php]: Ok, but where can i find an example to get all accessed devices at once?
I don't want to check every device in a single query for access.

Comment (Jellyfrog) [reply includes/html/print-dependency-map.inc.php]: Like so;
```Php
$device_ids = Permissions::devicesForUser()->toArray() ?: [0];
$sql = "WHERE `device_id` IN " .dbGenPlaceholders(count($device_ids));
$param = $device_ids;
```

However like I said in the other comment, this page should be written in laravel, and use the eloquent models functions

**Comment by SourceDoctor** [comment]
> This page could be written with blade/laravel, see #10559 for an example

@Jellyfrog 
i need some input from you to get rewrite done

i created a class file in directory `app/Http/Controllers`
there to get this 1:1 adapted in need following functions from `includes/common.php`:
- shorthost
- format_hostname
- generate_device_link

so i insert following lines into class file:
> use function includes\common\shorthost;
> use function includes\common\format_hostname;
> use function includes\common\generate_device_link;

but functions are still not found ...

any ideas?

**Comment by Jellyfrog** [comment]
You can do like `{!! \LibreNMS\Util\Url::deviceLink($device, $device->shortDisplayName()) !!}`

**Comment by SourceDoctor** [comment]
@Jellyfrog 
rewritten to blade/laravel
but still one problem i didn't get:
In Code everything is present, but on mouseover over a node only hostname is listed but it should short device information with some RRD Graphs be seen like on other maps. 
Can't find the cause of this.

File: app/Http/Controllers/MapController.php [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    {
+        return array();
+    }
+
+    // Device Dependency Map
+    public function dependencyMap(Request $request)
+    {
+        $user = $request->user();
+
+        if (!$user->hasGlobalRead()) {
```
Comment (Jellyfrog) [inline app/Http/Controllers/MapController.php]: You can replace this entire if with just `Device::hasAccess(Auth::user())`

Comment (Jellyfrog) [reply app/Http/Controllers/MapController.php]: Also add `->with('parents')` to eagerload all parents, saves a lot of queries.

Comment (SourceDoctor) [reply app/Http/Controllers/MapController.php]: both fixed

**Comment by SourceDoctor** [comment]
somebody an idea about the remaining mouseover Issue?

**Comment by SourceDoctor** [comment]
@Jellyfrog @PipoCanaja 
Travis test corrupted? 

> ...
> pylint requires Python '>=3.5.*' but the running Python is 2.7.12
> ...

**Comment by SourceDoctor** [comment]
ok, mouseover information runs.
But one problems remaining. Mouseover on every device type show mouseover display like it's a 'server'
even on switches PDU's and other device types

File: app/Http/Controllers/MapController.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+use App\Models\Device;
+use Illuminate\Http\Request;
+use LibreNMS\Config;
+use LibreNMS\Util\Url;
+
+class MapController extends Controller
+{
+    public function __construct()
+    {
+        $this->middleware('deny-demo');
```
Comment (Jellyfrog) [inline app/Http/Controllers/MapController.php]: this should be removed?

File: app/Http/Controllers/MapController.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    protected function visOptions()
+    {
+        return Config::get('network_map_vis_options');
+    }
+
+    protected function nodeDisabledStyle()
+    {
+        return array('color' => array(
```
Comment (Jellyfrog) [inline app/Http/Controllers/MapController.php]: you have a mix between `array()`and `[]`, stick to the newer `[]` style

File: app/Http/Controllers/MapController.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        $devices = Device::hasAccess($request->user())->with('parents')->get();
+
+        $dependencies = array();
+        $devices_by_id  = [];
+
+        // Build the style variables we need
+
+        // List all devices
+        foreach ($devices as $items) {
```
Comment (Jellyfrog) [inline app/Http/Controllers/MapController.php]: maybe change it to '$device`

File: app/Http/Controllers/MapController.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
+<?php
```
Comment (Jellyfrog) [inline app/Http/Controllers/MapController.php]: move this file into its own namespace and folder

File: app/Http/Controllers/MapController.php (hunk 3 above)
Comment (SourceDoctor) [reply app/Http/Controllers/MapController.php]: done

File: app/Http/Controllers/MapController.php (hunk 4 above)
Comment (SourceDoctor) [reply app/Http/Controllers/MapController.php]: done

File: app/Http/Controllers/MapController.php (hunk 5 above)
Comment (SourceDoctor) [reply app/Http/Controllers/MapController.php]: done

File: app/Http/Controllers/MapController.php (hunk 6 above)
Comment (SourceDoctor) [reply app/Http/Controllers/MapController.php]: @Jellyfrog 
my intention was to create a base for all Maps, so own directory is not necessary and doubled code can be prevented.
Especially for rewriting NetworkMap Code, it only has to include another method into MapController Class

what do you think about this?

Comment (Jellyfrog) [reply app/Http/Controllers/MapController.php]: I mean create a folder in Controllers named Maps and there u put all mapcontrollers. Check widget folder for inspiration

Comment (SourceDoctor) [reply app/Http/Controllers/MapController.php]: changed :-)

**Comment by SourceDoctor** [comment]
@Jellyfrog all changes are done

**Comment by SourceDoctor** [comment]
@Jellyfrog something else to change?

File: app/Http/Controllers/Maps/DependencyController.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                    'id'    => $device['device_id'],
+                    'label' => $device->shortDisplayName(),
+                    'title' => Url::deviceLink($device, null, [], 0, 0, 0, 0),
+                    'shape' => 'box',
+                ],
+                $device_style
+            );
+
+            // List all Device Dependencies
+            $parents = $device->parents()->get();
```
Comment (Jellyfrog) [inline app/Http/Controllers/Maps/DependencyController.php]: You're making N+1 queries here, and not using the eager cache.  
```suggestion
            $parents = $device->parents;
```

**Comment by Jellyfrog** [comment]
Probably it should be renamed DeviceDependencyController, since there are a lot of other dependencies

File: app/Http/Controllers/Maps/DependencyController.php (hunk 7 above)
Comment (SourceDoctor) [reply app/Http/Controllers/Maps/DependencyController.php]: changed

**Comment by SourceDoctor** [comment]
renamed

**Comment by Jellyfrog** [comment]
need to rename route and blade file also

**Comment by Jellyfrog** [comment]
need to rename route and blade file also

**Comment by SourceDoctor** [comment]
@Jellyfrog done

**Comment by djamp42** [comment]
I tried this out and because i have 6k devices it's unable to render the map, i left it for a couple of minutes, even if it does its probably not the best way to display this. I was hoping this would let me set the root node to start the dependency map drawing at.  Breaking them down by device group would be even better. It's a good start and i think graphing the dependency map is a really helpful and powerful tool.

**Comment by SourceDoctor** [comment]
@djamp42 
do you have the same issue with `Overview -> Maps ->Network`, cause Code is similar.

**Comment by djamp42** [comment]
> do you have the same problem with `Overview -> Maps ->Network`, because Code is similar.

Yes that page fails to load also, to be honest i never use it because seeing that many devices on one map doesn't make much sense. i only use the maps -> Device group maps.

File: app/Http/Controllers/Maps/DeviceDependencyController.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+use App\Models\Device;
+use Illuminate\Http\Request;
+use LibreNMS\Util\Url;
+
+class DeviceDependencyController extends MapController
+{
+    // Device Dependency Map
+    public function dependencyMap(Request $request)
+    {
+        $devices = Device::hasAccess($request->user())->with('parents')->get();
```
Comment (Jellyfrog) [inline app/Http/Controllers/Maps/DeviceDependencyController.php]: More eager caching
```suggestion
        $devices = Device::hasAccess($request->user())->with(['parents', 'location'])->get();
```

File: app/Http/Controllers/Maps/DeviceDependencyController.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+        $dependencies = [];
+        $devices_by_id  = [];
+
+        // Build the style variables we need
+
+        // List all devices
+        foreach ($devices as $device) {
+            if ($device['disabled']) {
```
Comment (Jellyfrog) [inline app/Http/Controllers/Maps/DeviceDependencyController.php]: `$device` is an object, please use `$device->disabled`

File: app/Http/Controllers/Maps/DeviceDependencyController.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+                $device_style = $this->nodeDisabledStyle();
+            } elseif (! $device['status']) {
+                $device_style = $this->nodeDownStyle();
+            } else {
+                $device_style = $this->nodeUpStyle();
+            }
+
+            // List all Device
+            $devices_by_id[$device['device_id']] = array_merge(
```
Comment (Jellyfrog) [inline app/Http/Controllers/Maps/DeviceDependencyController.php]: Why do you need to set array key to device id when you remove it with array_values?

File: resources/views/map/device-dependency.blade.php [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+@if($node_count)
+<div id="visualization"></div>
+@else
+<div class="alert alert-success" role="alert">@lang('No devices found')</div>
+@endif
+@endsection
+
+@section('scripts')
+<script src="js/vis.min.js"></script>
```
Comment (Jellyfrog) [inline resources/views/map/device-dependency.blade.php]: This line should be in `@section('javascript')`

Comment (Jellyfrog) [reply resources/views/map/device-dependency.blade.php]: And also use {{ asset('js/vis.min.js') }}

**Comment by Jellyfrog** [comment]
> > do you have the same problem with `Overview -> Maps ->Network`, because Code is similar.
> 
> Yes that page fails to load also, to be honest i never use it because seeing that many devices on one map doesn't make much sense. i only use the maps -> Device group maps.
> 
> It would seem to me this map should also work this way, where your building the dependency map based on the devices the device group returns. I took a quick look at the code and it seems it wouldn't be too hard to get this working, I'll have to play around with it more.

The problem is the js lib getting slow with many nodes.  
@SourceDoctor maybe you can look into updating it / changing some parameters?

**Comment by SourceDoctor** [comment]
I took before 2 weeks a look on project Page where vis.min.js comes from, about values and features. There seems to be a default limit to 600 Nodes. Have to play around with it some times later, especially about Engine Power.
Major Reason was lack of Feature to implement a Legend which describes Nodes and Connection States and Colors.
But this will be done from my side in another PR.
Also i'll implement here in another PR Node Highlighting. But first Code in #10943 has to be approved, also this PR.

File: app/Http/Controllers/Maps/DeviceDependencyController.php (hunk 8 above)
Comment (SourceDoctor) [reply app/Http/Controllers/Maps/DeviceDependencyController.php]: @Jellyfrog did i miss something?
i found in code nowhere usage of "location"

Comment (SourceDoctor) [reply app/Http/Controllers/Maps/DeviceDependencyController.php]: changed

File: app/Http/Controllers/Maps/DeviceDependencyController.php (hunk 9 above)
Comment (SourceDoctor) [reply app/Http/Controllers/Maps/DeviceDependencyController.php]: changed

File: app/Http/Controllers/Maps/DeviceDependencyController.php (hunk 10 above)
Comment (SourceDoctor) [reply app/Http/Controllers/Maps/DeviceDependencyController.php]: It's an adaption of `Map->Network`.
I try it without, run's also fine. So i removed array key setting.

File: resources/views/map/device-dependency.blade.php (hunk 11 above)
Comment (SourceDoctor) [reply resources/views/map/device-dependency.blade.php]: changed to `{{ asset('js/vis.min.js') }}`
changing to `@section('javascript')` has the effect no Map is graphed

Comment (Jellyfrog) [reply resources/views/map/device-dependency.blade.php]:
```blade
@section('javascript')
<script type="text/javascript" src="{{ asset('js/vis.min.js') }}"></script>
@endsection
```
Works for me.

Comment (SourceDoctor) [reply resources/views/map/device-dependency.blade.php]: ok,. this works on me also.
Changed to it

**Comment by SourceDoctor** [comment]
@djamp42 i implemented Device Dependency Graphing for Device Groups.
Hope it helps

**Comment by SourceDoctor** [comment]
@Jellyfrog for me it's ready for Merge

**Comment by SourceDoctor** [comment]
now implemented:

- redesign to blade/laravel
- filtering for Device Groups
- Device Highlighting

@Jellyfrog not sure if still something to do here (after 30 commits now ;) )
for my view it's ready to merge

**Comment by djamp42** [comment]
@SourceDoctor yeah i just tried this, and it's freaking awesome!! thanks alot!!!  Been playing around with it and it works perfectly fine with device groups as far as i can see.

File: app/Http/Controllers/Maps/DeviceDependencyController.php [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+                    $device_parents[] = $parent->device_id;
+            };
+
+            $childs = $device->children;
+            foreach ($childs as $child) {
+                    $device_childs[] = $child->device_id;
+            };
+        }
+
+        $devices_to_show = array_merge($device_nodes, array_merge($device_parents, $device_childs));
```
Comment (Jellyfrog) [inline app/Http/Controllers/Maps/DeviceDependencyController.php]: This selects \*ALL* devices and then filters them with a N+1 query.  
You could probably replace the entire foreach with:
```php
$devices = Device::inDeviceGroup($group_id)
  ->hasAccess($request->user())
  ->with([
    'location',
    'parents' => function ($query) use ($request) {
      $query->hasAccess($request->user());
    },
    'children' => function ($query) use ($request) {
      $query->hasAccess($request->user());
    }])
  ->get();

$devices = $devices->merge($devices->map->only('children', 'parents')->flatten());
``` 
If I'm not missing anything.

File: routes/web.php [hunk 13]
```diff
@@ -29,6 +29,13 @@
     Route::get('about', 'AboutController@index');
     Route::get('authlog', 'UserController@authlog');
 
+    // Maps
+    Route::group(['prefix' => 'maps', 'namespace' => 'Maps'], function () {
+        Route::get('devicedependency', 'DeviceDependencyController@dependencyMap');
+        Route::get('devicedependency/group={group_id}', 'DeviceDependencyController@dependencyMap');
```
Comment (Jellyfrog) [inline routes/web.php]: Lets either use query strings for this, or maybe better, create a new function in your controller to handle this.  
One for normal dependency, one for group, and one function that renders the map.
```php
Route::get('devicedependency/group/{group}', 'DeviceDependencyController@groupDependency');
```

```php
function groupDependency(Request $request, DeviceGroup $group) {
  // get devices...
  return self::renderMap($devices);
}
```

File: routes/web.php [hunk 14]
```diff
@@ -29,6 +29,13 @@
...
+        Route::get('devicedependency/group={group_id}/highlight_node={highlight_node}', 'DeviceDependencyController@dependencyMap');
```
Comment (Jellyfrog) [inline routes/web.php]: This should be removed and use `$request->get('highlight_node')` instead

Comment (SourceDoctor) [reply routes/web.php]: @Jellyfrog 
tried to remove both lines with variables
and in `DeviceDependencyController` changed 

```
public function dependencyMap(Request $request, $group_id = 0, $highlight_node = 0)
{
```
to
```
    public function dependencyMap(Request $request)
    {
    $group_id = $request->get('group_id') | 0;
    $highlight_node = $request->get('highlight_node') | 0;
```
but with this everytime GET arguments are in URL Page fall's back to Main Page

some Ideas?

Comment (SourceDoctor) [reply routes/web.php]: @Jellyfrog 
Code changed

File: app/Http/Controllers/Maps/DeviceDependencyController.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (SourceDoctor) [reply app/Http/Controllers/Maps/DeviceDependencyController.php]: changed

File: routes/web.php (hunk 13 above)
Comment (SourceDoctor) [reply routes/web.php]: group and highlight_node are now classical get Parameter

File: app/Http/Controllers/Maps/DeviceDependencyController.php [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+use App\Models\Device;
+use Illuminate\Http\Request;
+use LibreNMS\Util\Url;
+
+class DeviceDependencyController extends MapController
+{
+    protected static function deviceList($request) {
+
+        $group_id = $request->get('group') | 0;
```
Comment (Jellyfrog) [inline app/Http/Controllers/Maps/DeviceDependencyController.php]: `|` is a bitwise operator, you should use `||` if you want an `or`.   
In this case tho, You could just remove it

Comment (SourceDoctor) [reply app/Http/Controllers/Maps/DeviceDependencyController.php]: oh, yes you're right. :-)
i changed it to "||"

**Comment by SourceDoctor** [comment]
@Jellyfrog better now? :-)

File: app/Http/Controllers/Maps/DeviceDependencyController.php [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+                'location',
+                'parents' => function ($query) use ($request) {
+                    $query->hasAccess($request->user());
+                },
+                'children' => function ($query) use ($request) {
+                    $query->hasAccess($request->user());
+                }])
+            ->get();
+
+        return $devices->merge($devices->map->only('children', 'parents')->flatten());
```
Comment (Jellyfrog) [inline app/Http/Controllers/Maps/DeviceDependencyController.php]: add `->loadMissing('parents', 'location');` to this

Comment (SourceDoctor) [reply app/Http/Controllers/Maps/DeviceDependencyController.php]: @Jellyfrog 
Don't know how you want me to implement this.
Everywhere and with every combination of `parent`, `location`, `childrens` some Devices or Relationgraphs are missing.

Comment (Jellyfrog) [reply app/Http/Controllers/Maps/DeviceDependencyController.php]: Sorry it needs to be added after so it loads the data missing after the merge

Comment (SourceDoctor) [reply app/Http/Controllers/Maps/DeviceDependencyController.php]: do you mean to change it to:
`return $devices->merge($devices->map->only('children', 'parents')->loadMissing('parents', 'location')->flatten());`
This doesn't work at me. It throws following Error:

> production.ERROR: Method Illuminate\Support\Collection::loadMissing does not exist. {"userId":2,"exception":"[object] (BadMethodCallException(code: 0): Method Illuminate\\Support\\Collection::loadMissing does not exist. at /opt/librenms/vendor/laravel/framework/src/Illuminate/Support/Traits/Macroable.php:104)
[stacktrace]
#0 /opt/librenms/app/Http/Controllers/Maps/DeviceDependencyController.php(54): Illuminate\\Support\\Collection->__call('loadMissing', Array)
#1 /opt/librenms/app/Http/Controllers/Maps/DeviceDependencyController.php(68): App\\Http\\Controllers\\Maps\\DeviceDependencyController::deviceList(Object(Illuminate\\Http\\Request))
#2 [internal function]: App\\Http\\Controllers\\Maps\\DeviceDependencyController->dependencyMap(Object(Illuminate\\Http\\Request))
#3 /opt/librenms/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): call_user_func_array(Array, Array)

Comment (SourceDoctor) [reply app/Http/Controllers/Maps/DeviceDependencyController.php]: @Jellyfrog 
i changed it now to:
`return $devices->merge($devices->map->only('children', 'parents')->flatten())->loadMissing('parents', 'location');`
Hope it's right now :-)

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/daily-sh-not-running-automatically/9377/7

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-60-release-changelog-january-2020/10863/1

**Pull request merged**